### PR TITLE
feat(integrations): VSTS Status Sync Inbound

### PR DESCRIPTION
**Pull request by lauryndbrown**

### CODE REVIEW

**Comment by lauryndbrown** [comment]
https://cl.ly/423z3q3d0x16

File: src/sentry/integrations/vsts/issues.py [hunk 1]
```diff
@@ -156,3 +157,28 @@ def sync_status_outbound(self, external_issue, is_resolved, project_id, **kwargs
...
+        new_category = None
+        old_category = None
+        for state in all_states:
+            if state['name'] == new_state:
+                new_category = state['category']
+            if state['name'] == old_state:
+                old_category = state['category']
+
+        # what should I do if this doesn't work?
+        assert old_category is not None and new_category is not None
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/issues.py]: This is probably not the desired behavior. @macqueen Do you have suggestions?

File: src/sentry/integrations/vsts/webhooks.py [hunk 2]
```diff
@@ -33,6 +33,7 @@ def post(self, request, *args, **kwargs):
             try:
                 self.check_webhook_secret(request, integration)
             except AssertionError:
+                # TODO(lb): send a sentry error
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/webhooks.py]: I need to do this....

File: src/sentry/integrations/vsts/integration.py [hunk 3]
```diff
@@ -42,6 +42,12 @@
 
 class VstsIntegration(Integration, RepositoryMixin, VstsIssueSync):
     logger = logging.getLogger('sentry.integrations')
+    # TODO(lb): improvised until form changes are finalized.
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/integration.py]: Had to add these to get it to run correctly.

File: src/sentry/integrations/vsts/issues.py (hunk 1 above)
Comment (macqueen) [reply src/sentry/integrations/vsts/issues.py]: why wouldn't you just treat old/new category as being not "done" if it's`None`?

Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: @macqueen because if it's `None` something went wrong. Either the status is wrong or my code was wrong.  VSTS formats it like:
```
newValue: Active
oldValue: New
```
So there's not a possibility to ever have a `None` category unless something went wrong. 

Right now, though we aren't accounting for other types of workitems. I realize the change won't be very hard, but I wasn't aware that it was a priority. Let me know if so. Otherwise I think the above behavior is wrong, but I think we should at least log it. Not sure it's appropriate to act on invalid data...?

Comment (macqueen) [reply src/sentry/integrations/vsts/issues.py]: ok if it's not a scenario that should happen, i think it's fine to raise. I'm not sure I totally follow your last paragraph though -- when is the behavior wrong?

File: src/sentry/integrations/vsts/webhooks.py [hunk 4]
```diff
@@ -62,6 +70,29 @@ def handle_assign_to(self, integration, external_issue_key, assigned_to):
...
+        organization_ids = OrganizationIntegration.objects.filter(
+            integration_id=integration.id,
+        ).values_list('organization_id', flat=True)
+
+        old_category = None
+        new_category = None
+        for organization_id in organization_ids:
+            installation = integration.get_installation(organization_id)
+            if not (old_category and new_category):  # states should be the same throughout
+                old_category, new_category = installation.get_state_categories(
```
Comment (macqueen) [inline src/sentry/integrations/vsts/webhooks.py]: should we really be caching these across organizations? since in theory, the `default_auth_id` could be different for different orgs, it seems safer to not cache them. what if one org's default auth is no longer valid? i don't think we should use another org's auth to pull that data

File: src/sentry/integrations/vsts/issues.py (hunk 1 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: I'm sorry. Let me rephrase. I think an ` assert old_category is not None and new_category is not None` is wrong unless caught and given something appropriate to respond with. I assume we don't want a 500 error resulting here. 

I'm not sure about other webhooks, but VSTS actually will automatically disable a webhook if you have a number of unsuccessful responses. I've run into it several times while testing, so it can't be difficult to trigger. We don't currently handle that situation. 

That's my reasoning for not wanting an assertion error, but let me know if there's a different way I should think about it.

File: src/sentry/integrations/vsts/issues.py [hunk 5]
```diff
@@ -156,3 +157,28 @@ def sync_status_outbound(self, external_issue, is_resolved, project_id, **kwargs
...
+        return old_category, new_category
```
Comment (macqueen) [inline src/sentry/integrations/vsts/issues.py]: wondering if you could simplify this by doing something like:

```python
def get_done_states(self, project):
    if not self._done_states:
        client = self.get_client()
        all_states = client.get_work_item_states(self.instance, project)['value']
        self._done_states = [
            state['name'] for state in all_states if state['category'] in self.done_categories
        ]
    return self._done_states
```

then you could just use that in `should_resolve`/`should_unresolve`?  what do you think? just trying to think of ways to simplify/contain the logic since right now it's spread across this and the webhook

File: src/sentry/integrations/vsts/issues.py (hunk 1 above)
Comment (macqueen) [reply src/sentry/integrations/vsts/issues.py]: hmm i mean it seems like we want it to 500 in the case of "my code was wrong" if you're worried about the webhooks being disabled though, you could just log to sentry and then send a 400 response? up to you

Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: @macqueen I'm only worried because we don't have any way to handle this. And then the disabled webhook doesn't send any indication it's been disabled. The owner of the VSTS app just gets an email. So...? I'd prefer a 500, but unless the disabled webhook is handled, if anything goes wrong syncing will just stop happening.

File: src/sentry/integrations/vsts/webhooks.py (hunk 4 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/webhooks.py]: Gotcha!

File: src/sentry/integrations/vsts/webhooks.py [hunk 6]
```diff
@@ -62,6 +70,24 @@ def handle_assign_to(self, integration, external_issue_key, assigned_to):
...
+        ).values_list('organization_id', flat=True)
+
+        for organization_id in organization_ids:
+            installation = integration.get_installation(organization_id)
+            data = {
+                'new_state': status_change['newValue'],
+                'old_state': status_change['oldValue'],
+                'project': project,
+            }
+            installation.sync_status_inbound(external_issue_key, data)
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/webhooks.py]: @macqueen could adding `kwargs` to `sync_status_inbound` be a better answer or is that ok?

File: src/sentry/integrations/vsts/issues.py [hunk 7]
```diff
@@ -156,3 +157,19 @@ def sync_status_outbound(self, external_issue, is_resolved, project_id, **kwargs
...
+
+    def should_unresolve(self, data):
+        done_states = self.get_done_states(data['project'])
+        return data['old_state'] in done_states and not data['new_state'] in done_states
+
+    def should_resolve(self, data):
+        done_states = self.get_done_states(data['project'])
+        return not data['old_state'] in done_states and data['new_state'] in done_states
+
+    def get_done_states(self, project):
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/issues.py]: @macqueen Followed your advice on the structure. Let me know what you think.

File: src/sentry/integrations/vsts/integration.py [hunk 8]
```diff
@@ -74,8 +80,10 @@ def get_organization_config(self):
         instance = self.model.metadata['domain_name']
 
         try:
-            # NOTE(lb): vsts get workitem states does not give an id.
-            work_item_states = client.get_work_item_states(instance)['value']
+            # TODO(lb): this is clearly wrong. I'm just getting the first project.
```
Comment (macqueen) [inline src/sentry/integrations/vsts/integration.py]: do you mind filing a jira ticket to follow up on this? just so we don't forget

**Review by macqueen — APPROVED** [review]
✨
